### PR TITLE
Allow building with multiple SSL backends, selecting which one to use at runtime 

### PR DESCRIPTION
**Pull request by dscho**
Git for Windows uses cURL and offers users the choice whether to use OpenSSL or Secure Channel for its HTTPS transport. Unfortunately, this is an install-time option (swapping out binaries because it is a build time option in cURL), when it would be so much better a user experience if this could be a runtime option.

The latest cURL user survey also lists the desire for runtime-configurable SSL backends (and I was not the only one asking for it).

So here is a start.

What it does is this: it moves OpenSSL's and Secure Channel's private data into a union, and then introduces a new SSL backend dubbed "multi" that enables both members of that union, and sets a set of function pointers upon `Curl_ssl_init()` to the function implementations of the backend configured via the new environment variable `CURL_SSL_BACKEND`. Supported values are "openssl" and "schannel" for the moment.

In the interest of efficiency, I would like to solicit reviews and advice at this stage, which is already usable as far as Git for Windows' needs are concerned. In particular, I would love to get feedback on these questions:

- Is this the way to go, to move backend-specific stuff into the newly-introduced `backend` union in `struct connectdata`? Or would another method of adjusting the backends' code be preferable?
- Would it be okay to start just with OpenSSL and Secure Channel for now (because that is what Git for Windows needs)? As in: could this Pull Request go forward without converting the remaining SSL backends just yet (I am prepared to work on this in the future, of course).
- For the moment, I `#undef HAS_ALPN` before including the Secure Channel code in the multi backend. That's because my setup (a slightly modified MSYS2 environment) seems to offer ALPN support with OpenSSL, but not with Secure Channel. This would need to be a runtime option, right? Would it be okay to fold that into the newly-introduced `curlssl_features` enum? What other flags need to go there (I guess `HTTPS_PROXY_SUPPORT`, at least)?
- Each SSL backend seems to have a ton of `#define curlssl_*` at the end of its header file, offering  backend-specific functions via a unified interface. Should I introduce a new C++-like "virtual table" struct for these functions instead of setting function pointers as I do right now in `multi_init()`?
- How should I deal with the backend-specific dynamic discovery code in `configure`? I do not want to duplicate it (`--with-openssl` would need to perform the same auto-detection of files and features as `--with-multissl=openssl[...]`)...
- What API would you suggest to configure this from C code (I do not want to set the environment variable `CURL_SSL_BACKEND` in the caller, that looks a little too fragile to me)?

I would love to get lots of good feedback how to develop this topic branch so that it can eventually be integrated into cURL.

### CODE REVIEW

**Comment by dscho** [comment]
Oh, and any idea why Travis [fails](https://travis-ci.org/curl/curl/jobs/245355661#L2648) with:

```
make[2]: *** No rule to make target `../../src/libcurltool.la', needed by `unit1300'.  Stop.
make[2]: Leaving directory `/home/travis/build/curl/curl/tests/unit'
```

**Comment by MarcelRaad** [comment]
> Oh, and any idea why Travis fails with:

This is because checksrc.pl fails:
> 1804./getinfo.c:394:87: warning: Longer than 79 columns (LONGLINE)

... and some more. On Windows, you can run projects\checksrc.bat.

**Comment by bagder** [comment]
> Is this the way to go, to move backend-specific stuff into the newly-introduced backend union in struct connectdata? Or would another method of adjusting the backends' code be preferable?

Sorry, I haven't yet looked at the code so I'll just answer in general terms first: connectdata is the struct for each independent connection so I think it makes sense to make sure that knows which SSL backend is being used for that. I don't think we'll be able to mix/change SSL-backend on an existing connection so if you want to change, you need a new connection. (As a side-note to remember: this needs to be taken into account when re-using connections in multi-ssl mode.)

> Would it be okay to start just with OpenSSL and Secure Channel

Sure. But as you say we should probably consider how to make it work with other combos later on. And while the pandora is already out of the box: having more than two...

> For the moment, I #undef HAS_ALPN

Yeah, some of the backend build-time stuff clearly needs to be made into run-time conditionals since I suppose there will be a few features that not all backends in the multi-ssl bundle support uniformly!

> Each SSL backend seems to have a ton of #define curlssl_* at the end of its header file, offering backend-specific functions via a unified interface. Should I introduce a new C++-like "virtual table" struct for these functions

Yes, sounds like the way forward. Each backend could setup their own structs, similar to protocols set up structs for each protocol. Then we'd end up with a struct pointer to use for doing backend-ssl operations and for multi-ssl we would simply need to be able to use more than one/change the pointer run-time.

> What API would you suggest to configure this from C code

Tricky stuff. Can we start with you saying something how you envision this support to be used by application code? With a better understanding for the use case, maybe figuring out the API for it will be easier.

> How should I deal with the backend-specific dynamic discovery code in configure

From a user's perspective, I would imagine that just adding SSL-options to configure would be the logical way. Ie *both* `--with-ssl` and `--with-winssl` implies multi-ssl...

**Comment by dscho** [comment]
> > > Oh, and any idea why Travis fails with:
> 
> > This is because checksrc.pl fails:
> 
>     1804./getinfo.c:394:87: warning: Longer than 79 columns (LONGLINE)

Ah, thanks! That is a problem, of course, as the `<fieldname>` -> `backend.openssl.<fieldname>` type of changes require a lot of rewrapping.

Should I use a shorter name? Or introduce a macro? If so, what name would you prefer?

**Comment by dscho** [comment]
> I don't think we'll be able to mix/change SSL-backend on an existing connection

Oh... I did not mean to mix SSL backends, sorry, that use case never crossed my mind...

> as you say we should probably consider how to make it work with other combos later on. And while the pandora is already out of the box: having more than two...

That is easy: just add the straight-forward support for another SSL backend to `multissl.c` by `#include`'ing the source, `#undef`ing the `curlssl_*` constants, and extending `multi_init()` to handle the new set of functions.

It will be even easier with the vtable-like approach.

> > For the moment, I #undef HAS_ALPN
> 
> Yeah, some of the backend build-time stuff clearly needs to be made into run-time conditionals since I suppose there will be a few features that not all backends in the multi-ssl bundle support uniformly!

Right. The first commit already converts quite a few, such as `have_curlssl_certinfo`. So `HAS_ALPN` would need to be converted, too. Any other flags you can think off?

Related: I saw that some of the SSL backends' headers define quite a few constants that seem to be used only by the SSL backends themselves, e.g. `UNISP_NAME` in `schannel.h`, which is only used in `schannel.c`. Is that intentional? Isn't this bleeding implementation details, i.e. shouldn't those definitions be moved into the respective `.c` files?

> > Each SSL backend seems to have a ton of #define curlssl_* at the end of its header file, offering backend-specific functions via a unified interface. Should I introduce a new C++-like "virtual table" struct for these functions
> 
> Yes, sounds like the way forward. Each backend could setup their own structs, similar to protocols set up structs for each protocol. Then we'd end up with a struct pointer to use for doing backend-ssl operations and for multi-ssl we would simply need to be able to use more than one/change the pointer run-time.

You mean `struct Curl_handler`, where e.g. FTP is supported using [`struct Curl_handler Curl_handler_ftp`](https://github.com/curl/curl/blob/b4043706eba8436cf82815277af9153c687757c7/lib/ftp.c#L162-L185)?

That sounds really clean to me, too.

> > What API would you suggest to configure this from C code
> 
> Tricky stuff. Can we start with you saying something how you envision this support to be used by application code? With a better understanding for the use case, maybe figuring out the API for it will be easier.

Oh, sorry, should have said that earlier.

The use case is `git clone`/`git fetch`/`git push` via HTTPS. It is quite common in enterprise to use custom certificates that are distributed via enterprise management tools, and every developer's machine's Windows Certificate Store will have them. That requires Secure Channel, as OpenSSL (which Git for Windows used for ages) does not know how to access the Windows Certificate Store.

So what I imagine is that either the administrator or the user set a config variable (say, `http.sslbackend`) that Git will then use to configure cURL prior to every clone/fetch/push operation.

In other words, it needs to be set *once* per process. Git for Windows would not need the ability to re-configure cURL to use another one after the initial configuration.

> > How should I deal with the backend-specific dynamic discovery code in configure
> 
> From a user's perspective, I would imagine that just adding SSL-options to configure would be the logical way. Ie both --with-ssl and --with-winssl implies multi-ssl...

That's a good idea! I just do not know how to handle the default SSL backend, then (you'd want to use the first configured one, right?).

And that does not even start to handle CMake, with which I ask for a lot more guidance because I know it so little...

**Comment by dscho** [comment]
Oh, about the `HAS_ALPN` issue: this is a non-issue, really, and only a sign that I did things a bit *too* hacky: the `HAS_ALPN` constant is only ever defined and used in individual `.c` files. The problem I faced arose only because I `#include`d `.c` files...

**Comment by dscho** [comment]
Okay, buckle up...

**Comment by dscho** [comment]
This is an almost complete rewrite of the topic branch. And it is quite a bit bigger. I tried my best to reorder the patches in a logical manner, and to split the changes into meaningful commits.

This time, I handle *all* SSL backends (although you have to configure at build time which ones you want to build against explicitly).

This time, `./configure --with-ssl --with-winssl --with-gnutls` works as one would expect.

Since I do not have access to all of the SSL backends (only GNU TLS, OpenSSL and Secure Channel), I could not compile-test, unfortunately. The good news is that in practice, it will only matter if anybody tries to build with a different set of multiple backends: in that case, something like https://github.com/curl/curl/pull/1601/commits/65c442eab2f06d3e6fb9bb768e98e5e7aa49271b may be needed (i.e. some `#undef` tricks in case of clashing definitions in `#include`d header files).

As before, though, the runtime configuration happens via the environment variable `CURL_SSL_BACKEND` (the easiest test that it worked, I found, is to run `CURL_SSL_BACKEND=openssl ./src/curl --version` and look at the right part of the first line). @bagder I still would love to hear your ideas how to improve that in a compile-time safe manner.

Also as before, CMake is not handled at all. But then, it looks as if CMake lags behind on SSL backends anyway :smile: 

Also as before, I have no test coverage (I have not mentioned this before, though). Any guidance?

**Comment by dscho** [comment]
Oh, i just realized that GitHub juggled the commits around in the PR view, because it orders them by commit date instead of topologically. Please let me know if you prefer strongly for me to rewrite the commits so that their commit dates do not reflect the original commits' dates, but are completely rewritten instead to force GitHub to show them in the correct order.

**Comment by dscho** [comment]
Any way I can help move this forward?

**Comment by bagder** [comment]
Sorry, it's just me being a bit slow. I haven't forgot about it and I don't mean to intentionally leave you behind. I'll shape up and get some feedback to you on this soonish!

**Comment by dscho** [comment]
@bagder no worries! You had a tough week, I fully understand. Besides, you are an OSS maintainer, and I understand the pressures of that role, too.

For the record: I would love to get this into Git for Windows soon-ish, even as an experimental, patched version of cURL. To that end, I want to wait a little for the feedback I get, to gage how much of the design needs to change (if only at most cosmetic changes are necessary to get this merged, I would be happy with taking an early version for an experimental/intermediate MinGit for Windows).

So if there is a chance for an early-bird preview of an overall verdict ("design looks good" or "actually, the way this is done needs to be overhauled to use XYZ instead of ABC"), I would be most thankful.

**Comment by bagder** [comment]
I think the general design look s perfectly fine!

I see you opted to make it possible to change ssl backend by setting an environment variable. Is that how you actually would like it done? For general applications that's a pretty awkward API so there should at least also be an additional option to set somehow, and possibly one of those options could be one that allows the environment variable control.

There are compiler warnings in vtls/vtls.c (I recommend configure's `--enable-werror` option, and I plan to add it to some CI build soon to make it detect warnings better). Mostly harmless "unused parameter" ones.

Nit: please don't use `Curl_multi_` as a prefix, since that's already used a little for functions in lib/multi.c. I suggest maybe `Curl_mssl_` or `Curl_multissl_` (although the latter one is a bit on the long side).

**Comment by dscho** [comment]
> I think the general design looks perfectly fine!

Thanks! :smile: 

> I see you opted to make it possible to change ssl backend by setting an environment variable. Is that how you actually would like it done? For general applications that's a pretty awkward API so there should at least also be an additional option to set somehow, and possibly one of those options could be one that allows the environment variable control.

I find the environment variable less than ideal myself. However, I have no idea what the proper way is, according to cURL style.

Is there a common way to pre-configure certain defaults before initializing cURL? Or should we have it as an option to be configured via the "curl_easy", safeguarding against setting it multiple times?

> There are compiler warnings in vtls/vtls.c (I recommend configure's --enable-werror option, and I plan to add it to some CI build soon to make it detect warnings better). Mostly harmless "unused parameter" ones.

I tried to compile this with `--enable-werror --with-ssl --with-gnutls` in my Linux VM (GCC 5.4.0) and with `--enable-werror --with-ssl --with-gnutls --with-winssl` on Windows (GCC 6.3.0), and I did not get any warnings. Would you mind pasting the compile output (or at least the relevant part thereof)?

> Nit: please don't use Curl_multi_ as a prefix, since that's already used a little for functions in lib/multi.c. I suggest maybe Curl_mssl_ or Curl_multissl_ (although the latter one is a bit on the long side).

Okay, fair enough. I am not too happy about "multissl" either. I thought of "varssl", but it still does not convey the "configure at run-time" nature of the backend... So far, I am still leaning toward "multissl".

**Comment by bagder** [comment]
> Is there a common way to pre-configure certain defaults before initializing cURL?

I've struggled with figuring out the best way for this init.

There is the `curl_global_init()` function that is one that triggers the TLS lib init today, but it only has "flags" argument to add stuff to, and that is probably too limited to ask for multi-ssl-init? (or perhaps it'll work out if we reserve one bit for each known TLS lib?)

Otherwise I've been thinking that since multi-ssl is new functionality anyway, we might want to consider adding a totally new way to init this as well.

> Or should we have it as an option to be configured via the "curl_easy", safeguarding against setting it multiple times?

That would be cool, but several of the TLS libraries have thread-unsafe inits, so we want them done at the time of the global_init so that applications can have more control of that specific phase.

## warnings

Sorry for being brief, I just rebuilt the code with my "usual" configure setup which boils down to:

`./configure --disable-shared --enable-debug --enable-maintainer-mode --enable-werror`

..  this finds and uses OpenSSL 1.1.0f (only). That then generates this set of warnings:

```
vtls/vtls.c: In function ‘Curl_none_shutdown’:
vtls/vtls.c:968:44: error: unused parameter ‘conn’ [-Werror=unused-parameter]
 int Curl_none_shutdown(struct connectdata *conn, int sockindex)
                                            ^~~~
vtls/vtls.c:968:54: error: unused parameter ‘sockindex’ [-Werror=unused-parameter]
 int Curl_none_shutdown(struct connectdata *conn, int sockindex)
                                                      ^~~~~~~~~
vtls/vtls.c: In function ‘Curl_none_check_cxn’:
vtls/vtls.c:973:45: error: unused parameter ‘conn’ [-Werror=unused-parameter]
 int Curl_none_check_cxn(struct connectdata *conn)
                                             ^~~~
vtls/vtls.c: In function ‘Curl_none_random’:
vtls/vtls.c:978:45: error: unused parameter ‘data’ [-Werror=unused-parameter]
 CURLcode Curl_none_random(struct Curl_easy *data, unsigned char *entropy,
                                             ^~~~
vtls/vtls.c:978:66: error: unused parameter ‘entropy’ [-Werror=unused-parameter]
 CURLcode Curl_none_random(struct Curl_easy *data, unsigned char *entropy,
                                                                  ^~~~~~~
vtls/vtls.c:979:34: error: unused parameter ‘length’ [-Werror=unused-parameter]
                           size_t length)
                                  ^~~~~~
vtls/vtls.c: In function ‘Curl_none_close_all’:
vtls/vtls.c:984:44: error: unused parameter ‘data’ [-Werror=unused-parameter]
 void Curl_none_close_all(struct Curl_easy *data)
                                            ^~~~
vtls/vtls.c: In function ‘Curl_none_session_free’:
vtls/vtls.c:987:35: error: unused parameter ‘ptr’ [-Werror=unused-parameter]
 void Curl_none_session_free(void *ptr)
                                   ^~~
vtls/vtls.c: In function ‘Curl_none_data_pending’:
vtls/vtls.c:990:55: error: unused parameter ‘conn’ [-Werror=unused-parameter]
 bool Curl_none_data_pending(const struct connectdata *conn, int connindex)
                                                       ^~~~
vtls/vtls.c:990:65: error: unused parameter ‘connindex’ [-Werror=unused-parameter]
 bool Curl_none_data_pending(const struct connectdata *conn, int connindex)
                                                                 ^~~~~~~~~
vtls/vtls.c: In function ‘Curl_none_set_engine’:
vtls/vtls.c:1000:49: error: unused parameter ‘data’ [-Werror=unused-parameter]
 CURLcode Curl_none_set_engine(struct Curl_easy *data, const char *engine)
                                                 ^~~~
vtls/vtls.c:1000:67: error: unused parameter ‘engine’ [-Werror=unused-parameter]
 CURLcode Curl_none_set_engine(struct Curl_easy *data, const char *engine)
                                                                   ^~~~~~
vtls/vtls.c: In function ‘Curl_none_set_engine_default’:
vtls/vtls.c:1005:57: error: unused parameter ‘data’ [-Werror=unused-parameter]
 CURLcode Curl_none_set_engine_default(struct Curl_easy *data)
                                                         ^~~~
vtls/vtls.c: In function ‘Curl_none_engines_list’:
vtls/vtls.c:1010:61: error: unused parameter ‘data’ [-Werror=unused-parameter]
 struct curl_slist *Curl_none_engines_list(struct Curl_easy *data)
                                                             ^~~~
vtls/vtls.c: In function ‘Curl_none_md5sum’:
vtls/vtls.c:1021:57: error: unused parameter ‘md5len’ [-Werror=unused-parameter]
                           unsigned char *md5sum, size_t md5len)
                                                         ^~~~~~
vtls/vtls.c: At top level:
vtls/vtls.c:1071:1: error: ‘static’ is not at beginning of declaration [-Werror=old-style-declaration]
 const static struct Curl_ssl Curl_ssl_multi = {
 ^~~~~
vtls/vtls.c:1105:1: error: ‘static’ is not at beginning of declaration [-Werror=old-style-declaration]
 const static struct Curl_ssl *available_backends[] = {
 ^~~~~
cc1: all warnings being treated as errors

```

**Comment by bagder** [comment]
The warnings are actually also in the travis build (I think we should consider doing `--enable-errror` on those to detect this better):

See: https://travis-ci.org/curl/curl/jobs/247276853#L1360

**Comment by bagder** [comment]
Oops, yes, we do werror already since fc2e81c38b =)

**Comment by dscho** [comment]
> > Is there a common way to pre-configure certain defaults before initializing cURL?
> 
> I've struggled with figuring out the best way for this init.
> 
> There is the curl_global_init() function that is one that triggers the TLS lib init today, but it only has "flags" argument to add stuff to, and that is probably too limited to ask for multi-ssl-init? (or perhaps it'll work out if we reserve one bit for each known TLS lib?)
> 
> Otherwise I've been thinking that since multi-ssl is new functionality anyway, we might want to consider adding a totally new way to init this as well.

So maybe a new function `curl_select_sslbackend(curl_sslbackend backend);` that specifically verifies that `curl_global_init()` has not yet run? And also errors out if support for the specified has not been compiled in?

> Sorry for being brief, I just rebuilt the code with my "usual" configure setup which boils down to:
> 
> ./configure --disable-shared --enable-debug --enable-maintainer-mode --enable-werror

Ah... got it...

> The warnings are actually also in the travis build

D'oh, I should have checked, and then simply read `.travis.yml`. Sorry for the noise. Will update this Pull Request and ping you.

**Comment by bagder** [comment]
> maybe a new function curl_select_sslbackend(curl_sslbackend backend); that specifically verifies that curl_global_init() has not yet run? And also errors out if support for the specified has not been compiled in?

(brainstorming mode, just writing out my ideas and thoughts here)

Maybe something like this (I picked the `curl_global_` prefix here since it is a global function and is similar in spirit to `curl_global_init`):

`CURLsslset curl_global_sslset(curl_sslbackend backend, curl_sslbackend *avail);
`

Returns:
- `CURLSET_OK` on success
- `CURLSET_UNKNOWN_BACKEND` for "unsupported backend" to select 
- `CURLSET_TOO_LATE` when getting called "too late, already inited"
- etc

The **avail** value is a bitmask that is returned with information about which SSL backends that are available. The bit number in the bitmask represents the particular backend: `1<<backend`.

That way an application can call `curl_global_sslset(CURLSSLBACKEND_NONE, &alternatives);` to query the library for what options there are.

**Comment by dscho** [comment]
> Maybe something like this (I picked the curl_global_ prefix here since it is a global function and is similar in spirit to curl_global_init):
> 
> ```c
> CURLsslset curl_global_sslset(curl_sslbackend backend, unsigned int *avail);
> ```

Okay, the way I attempted this in my latest commit was to return `CURLcode` instead, but your approach is much cleaner.

I am not that sure about the bitfield, though... I guess it is safe for now, as we "only" have 11 different backends. But what to do with this bitfield? What if an application links to cURL 7.55.0 with multissl support, then cURL 7.56.0 introduces five new SSL backends and avail[18] is set? What would the application say, then... `SSL backend #18 is available`, maybe? That is not really user-friendly...

Additionally, the way I want to use this from Git for Windows is to have the user configure the backend. For that, I do not want to use IDs but labels, e.g. `"openssl"`.

That is why I introduced the `curl_select_sslbackend_by_name()` function, and I think our final version needs to have support for selecting the SSL backend by name rather than by ID.

I guess that
```c
CURLsslset curl_global_sslset(const char *name, char **avail);
CURLsslset curl_global_sslset_by_id(curl_sslbackend backend, unsigned int *avail);
```
would address this concern.

Another concern I have: if cURL was built with `USE_SSL` but there are no backends in `available_backends`, I print an error to `stderr` and `exit(1);`. That is certainly wrong for a library to do, but I was unsure how to react when `Curl_ssl_backend()` is called in such a scenario. Maybe simply punt and return -1, without writing anything to `stderr`?

BTW I just realized that one debug build tests specifically for the order of functions in the header. Is this needed e.g. to provide a consistent numbering in `.dll` files? Do I need to append the declarations at the end of `curl.h`?

**Comment by bagder** [comment]
All good remarks!

Right, returning an avail bitmask is pretty lame for when we introduce new backends that the application didn't know about when it was written. I think we should stick with providing one function for this purpose, so let's stick to strings I think. An extended thought there is if we should consider returning an array of structs for the backends and allow that struct to have some sort of capabilities field or something, like if an application would like to pick a backend that supports feature X...

And yes, no stderr use by default (but we have no way to switch it on by choice) and exit() is forbidden. I think a clear and documented return code for that case is good enough.

**Comment by dscho** [comment]
> An extended thought there is if we should consider returning an array of structs for the backends and allow that struct to have some sort of capabilities field or something, like if an application would like to pick a backend that supports feature X...

Like, return `available_backends` (which is `NULL` terminated) as a `const struct Curl_ssl *`? We would only need to export the declaration of that struct (and then keep it stable). i kinda like that.

We could even be stricter and declare part of `struct Curl_ssl` (the part about the name, the id and the capabilities) as public struct, make that the first element of a private struct, and then return an array of pointers "to the public struct".

> And yes, no stderr use by default (but we have no way to switch it on by choice) and exit() is forbidden. I think a clear and documented return code for that case is good enough.

Okay.

> BTW I just realized that one debug build tests specifically for the order of functions in the header. Is this needed e.g. to provide a consistent numbering in .dll files? Do I need to append the declarations at the end of curl.h?

What about his?

**Comment by bagder** [comment]
> We could even be stricter and declare part of struct Curl_ssl (the part about the name, the id and the capabilities) as public struct, make that the first element of a private struct, and then return an array of pointers "to the public struct".

Right, I think we should return a public struct that we only use for exporting info, we use our own internal ones separate. In the long run it is always better to keep them independent.

> I just realized that one debug build tests specifically for the order of functions in the header. Is this needed e.g. to provide a consistent numbering in .dll files? Do I need to append the declarations at the end of curl.h?

Right, that's because some architectures/builds require consistent ordering and they get the order from that header file. So new functions should be appended last in order to cause the least amount of headaches.

**Comment by dscho** [comment]
@bagder I hope you had a great vacation!

So here is the promised update, the relevant changes are in the last three commits: https://github.com/dscho/curl/compare/273a8dd69cb20dfda591be6f63306016c7d38c2f~3...273a8dd69cb20dfda591be6f63306016c7d38c2f.

**Comment by dscho** [comment]
So. I tried adding the new function as last one to `include/curl/curl.h`, but test 1135 still complains. And from what I see in `packages/OS400/make-lib.sh`, rightfully so, as *anything* in `curl.h` will come before anything from `easy.h`, `mprintf.h` and `multi.h`.

I could add the function declaration to `multi.h`, of course, but that would be super ugly.

I could also introduce a new header, say `include/curl/sslbackend.h`, but that also looks very inelegant.

And finally, it seems I could bump the SONAME, but that is *really* inelegant.

What to do???

*clicketyclick* *gitlog* *gitblame*

Ah.

It would appear that the HTTP/2 pushheader functions were appended to `multi.h` for the very same reason that test 1135 fails for this Pull Request...

So I will add the declaration to that header file, too, even if I find it ugly ;-)

**Comment by dscho** [comment]
Okay, I made it so, and for good measure I added a man page and an example.

The relevant changes relative to the latest reviewed iteration of this Pull Request are: https://github.com/dscho/curl/compare/e3c4314af~5...e3c4314af

File: docs/libcurl/symbols-in-versions [hunk 1]
```diff
@@ -689,6 +689,9 @@ CURLSSLBACKEND_SCHANNEL         7.34.0
 CURLSSLBACKEND_WOLFSSL          7.49.0
 CURLSSLOPT_ALLOW_BEAST          7.25.0
 CURLSSLOPT_NO_REVOKE            7.44.0
+CURLSSLSET_OK                   7.56.0
+CURLSSLSET_UNKNOWN_BACKEND      7.56.0
+CURLSSLSET_TOO_LATE             7.56.0
```
Comment (dscho) [inline docs/libcurl/symbols-in-versions]: I should also draw attention to these lines: it is assumption on my part that this PR could make it into cURL 7.56.0...

**Comment by dscho** [comment]
While reviewing the merge conflicts, I realized that the SSL backend-specific data should be further encapsulated, i.e. they should not be in a `union` with all the other backends' internal data.

I will work on that, but it could take a little while as I will be busy this week with another project.

Stay tuned!

**Comment by dscho** [comment]
I still would like to describe the `malloc()` alignment issues and how I solved them in the commit message, and I still want to try to verify axTLS, GSKit and PolarSSL. Otherwise, I think this should be pretty much the final state of this PR. What do you think, @bagder?

**Comment by dscho** [comment]
Okay, so i updated the commit message to clarify how `allocate_conn()` was changed and why, and I also tested axTLS (a minor change was required: `Curl_axtls_close()` had to be moved before its first caller).

I could not get my hands at GSKit, but at least I went through the source code trying to spot any obvious problems (use of functions before they are defined, cast of pointers to `const struct connectdata`'s fields to non-const, etc) and did not find any.

I finally looked at PolarSSL and saw that it is simply an obsolete name for mbedTLS, so I did not really bother trying to build with both PolarSSL and mbedTLS backends (I am not even sure whether that would work).

But I managed to build and test with `./configure --disable-shared --enable-debug --enable-maintainer-mode --enable-werror --with-ssl --with-gnutls --with-nss --with-mbedtls --with-cyassl --with-axtls=<custom-installation>` on Linux, and `./configure --disable-shared --enable-debug --enable-maintainer-mode --enable-werror --with-ssl --with-gnutls --with-winssl` on Windows (modulo the problems mentioned in https://github.com/curl/curl/pull/1652).

So as soon as CI agrees, I would call this PR ready to rock 'n roll.

**Comment by dscho** [comment]
> So as soon as CI agrees, I would call this PR ready to rock 'n roll.

Guess it is Rock 'n Roll time. @bagder what do you think?

**Comment by bagder** [comment]
I think it looks really good: 👍. I'm just rather occupied for the moment with bug fixes and getting all ducks in a row for the upcoming release. I see no major obstacles for merging this once that is out, so I'll get back on this after August 9.

**Comment by dscho** [comment]
Okay. Are you working on a .0 release, or a bug fix release?

**Comment by bagder** [comment]
curl 7.55.0 will ship on August 9, [Pending release notes for that release](https://curl.haxx.se/dev/release-notes.html).

**Comment by dscho** [comment]
> curl 7.55.0 will ship on August 9,

Thanks! So my change to `docs/libcurl/symbols-in-versions` (mentioning that the `CURLSSLSET_*` constants appear in 7.56.0) in e9283bf2530a10e0440d6c4b02fdb5ad6973394d was pretty prescient, eh?

**Comment by bagder** [comment]
Yeps, that has every chance of being correct! =)

**Comment by dscho** [comment]
Seems that the Travis build failed because of *apt problems*, i.e. it could not download some Ubuntu packages.

Is there any way to kick the build off again? I am fairly certain that this PR should be all green by now.

**Comment by MarcelRaad** [comment]
@dscho Thanks, I've restarted the failing job (Linux / GCC / normal).

**Comment by bagder** [comment]
Once the dust settles after the hiccups of 7.55.0 and the follow-up 7.55.1 release, I intend to give this series of commits my thorough review (with a particular focus on the external interface APIs and code that we can't easily change once we release it). Ideally we then discuss any nits/feedback I may have and then I proceed and merge into master, on its way to towards fame and glory in the future 7.56.0 release! :-)

It needs to be done by September 6 or it'll miss this release, but it feels like plenty enough time.

**Comment by dscho** [comment]
Awesome!

**Comment by bagder** [comment]
In general, very nice nice work. I think you've managed to work with our code style and general approach really well. A large piece to merge, but I feel really good about it. My compliments!

A few things to fix and consider before merge:
1. metalink build. I realized we don't build with metalink in the CI builds (fixed in ca9630f1284), but when I did the build breaks. The metalink code is outside of the library but yet uses crypto functions from the TLS libs - its a bit ugly, but needs some love to work with this overhaul.
2. the man page for curl_global_sslset needs an AVAILABILITY section that says it is introduced in 7.56.0
3. to consider: a curl_version_info feature flag for multi-ssl. Then for example `curl -V` should show it being multi-ssl enabled easily.
4. `curl_version` doesn't show *any* TLS backend and version anymore? It seems `curl -V` doesn't show any TLS and lib version now, which is a pity. Of course there's a question of what to show for numerous TLS libs but I think the most sensible thing is to include all that are built-in.
5. docs: if libcurl is built multi-ssl, and no backend is selected, which one is used? And also, is that default selection possible to select build-time? Since most users out there will not set backend when this gets introduced it is fairly important questions to users.

I've successfully tested building with various TLS combinations both on Linux (with OpenSSL, GnuTLS, NSS, axTLS and wolfssl) and Mac (withopenssl and darwinssl). 👍

**Comment by dscho** [comment]
> metalink build. I realized we don't build with metalink in the CI builds (fixed in ca9630f), but when I did the build breaks. The metalink code is outside of the library but yet uses crypto functions from the TLS libs - its a bit ugly, but needs some love to work with this overhaul.

Fixed in 8545c6d5b.

> the man page for curl_global_sslset needs an AVAILABILITY section that says it is introduced in 7.56.0

Fixed in 3490dfc06.

> to consider: a curl_version_info feature flag for multi-ssl. Then for example curl -V should show it being multi-ssl enabled easily.

Implemented in fc1653a65.

> curl_version doesn't show any TLS backend and version anymore? It seems curl -V doesn't show any TLS and lib version now, which is a pity. Of course there's a question of what to show for numerous TLS libs but I think the most sensible thing is to include all that are built-in.

Implemented in 8a0194404. I put all but the active SSL backends in parentheses.

> if libcurl is built multi-ssl, and no backend is selected, which one is used?

Heh... This is not possible, as multi-ssl is not a `./configure` option, it is *implicit* when you build with more than one backend.

In fact, it was implicit in the previous iterations even when you build with only one backend. It always went through `multissl_init()` to select the backend at runtime.

I found that inelegant, so I added a patch (8f47553f8) that configures the single available SSL backend already at build time if there is only one to begin with, without having to go through `multissl_init()`.

> And also, is that default selection possible to select build-time? Since most users out there will not set backend when this gets introduced it is fairly important questions to users.

No, it is currently not possible to select the default SSL backend at build-time, I punted on that one for now. There is the alphabetically-sorted array `available_backends`, and its first entry is automatically the default.

It should be relatively easy to change that in an add-on patch, but I am a bit uncertain about an elegant way to ensure at compile time that the chosen default is actually compiled in... If you have a splendid idea how to do that, I would really like to hear it.

**Comment by dscho** [comment]
Okay, i added two more commits (and rebased the whole shebang, sorry that my previous "fixed in <commit>" no longer apply verbatim).

Now you *can* set the default SSL backend via `./configure --with-default-ssl-backend=<backend>`. It s not compile-time safe, as I mentioned earlier, but the value is tested in `./configure` against the list of to-be-compiled-in SSL backends.

While working on that, I noticed something I had missed earlier: the symbols are optionally prefixed with the current SSL backend. If more than one SSL backend was compiled in, *one* of them was picked for said versioning. I changed that to the `MULTISSL_` prefix.

While at it, I also moved the determination whether more than one SSL backend was configured into the `./configure` script, as it needs that information for the versioning anyway. This may make things potentially a bit harder for folks interested in getting CMake up to par with multiple SSL backends.

@bagder I hope you agree with this course of action?

**Comment by dscho** [comment]
> It needs to be done by September 6 or it'll miss this release, but it feels like plenty enough time.

Is this PR still on track for 7.56.0?

**Comment by bagder** [comment]
absolutely!

**Comment by bagder** [comment]
As of commit a330bab353fb1f9, this is now merged into master.

Thanks @dscho for your hard work on this. Now let's see what bugs we have left to fix for this! =)

**Comment by dscho** [comment]
> Thanks @dscho for your hard work on this. Now let's see what bugs we have left to fix for this! =)

Awesome! Please ping me with whatever fallout there is to fix, and I will try my best to address the issues as quickly as time allows.

**Comment by dscho** [comment]
(And sorry about the incorrect version in the man page, I saw that you fixed this in `master`, thank you so much!)
